### PR TITLE
Update to sprockets 4

### LINKED ISSUE #1972: Sprockets is pinned to < 4
I'd like to update govuk-developer-docs to use the latest version of the govuk_tech_docs gem.

There's a version conflict between govuk_tech_docs and govuk_publishing_components.

* `govuk_tech_docs` [requires sprockets `~> 4.0.0`](https://github.com/alphagov/tech-docs-gem/blob/master/govuk_tech_docs.gemspec#L49)
* `govuk_publishing_components` [requires sprockets `< 4`](https://github.com/alphagov/govuk_publishing_components/blob/master/govuk_publishing_components.gemspec#L24)

This was deliberately pinned to `< 4` in #1172 because of some issue with CI. It's not clear whether this is still a problem, and if so whether it's something we could fix.

### PR DESCRIPTION
**Pull request by alex-ju**
## What
This PR updates sprockets to latest major versions, which require precompiled assets to be declared via `manifest.js`

## Why
To keep our dependencies up to date. Fixes #1972

## Notes
Configuration tested with:

- [x] collections
- [x] content-publisher
- [ ] email-alert-frontend
- [ ] feedback
- [x] finder-frontend
- [x] frontend
- [x] government-frontend
- [ ] govspeak
- [ ] info-frontend
- [ ] licence-finder
- [ ] manuals-frontend
- [ ] service-manual-frontend
- [x] smart-answers
- [ ] whitehall
- [x] static

## Visual Changes
n/a

### CODE REVIEW

**Comment by kevindew** [comment]
Hey @alex-ju, as discussed on Slack, I've also been working on this - and I'm sorry we've ended up both putting energy into the same thing (I wonder what force of nature inspired us both.

The branch I've got is: https://github.com/alphagov/govuk_publishing_components/tree/sprockets-4-kevin and you can see the diff in: https://github.com/alphagov/govuk_publishing_components/compare/sprockets-4-kevin

Probably the major difference between our approaches is that I took a stance of supporting both Sprockets 3 and 4 for a while and warning users to upgrade.

There's a few things that seem to be working in your branch that surprise me too. I'll leave some comments surrounding those. 

I'm not precious about my work on this so happy for you to make my bits or to open my own PR. Let me know your preference.

File: app/assets/config/govuk_publishing_components_manifest.js [hunk 1]
```diff
@@ -1,3 +0,0 @@
-// JS and CSS bundles for the gem
```
Comment (kevindew) [inline app/assets/config/govuk_publishing_components_manifest.js:1]: How come you changed the name of this file? I found having this was quite useful for adding it to apps so they had:

```
link govuk_publishing_components_manifest
```

but it'd be better still if it wasn't needed, does using manifest.js mean it gets pulled in automatically?
For what it's worth the Rails engine documentation suggests `<engine_name>_manifest.js`: https://guides.rubyonrails.org/engines.html#generating-an-engine

File: config/initializers/assets.rb [hunk 2]
```diff
@@ -52,3 +56,9 @@
   #{__dir__}/../../node_modules/govuk-frontend/
   #{__dir__}/../../node_modules/
 ]
+
+# Disable concurency otherwise the assets will be compiled in a random order
+# resulting in failures where, for example, a mixin is called but not yet compiled
+Rails.application.config.assets.configure do |environment|
+  environment.export_concurrent = false
+end
```
Comment (kevindew) [inline config/initializers/assets.rb:64]: Oh this is strange, I didn't encounter this.

File: govuk_publishing_components.gemspec [hunk 3]
```diff
@@ -22,7 +22,7 @@ Gem::Specification.new do |s|
   s.add_dependency "plek"
   s.add_dependency "rails", ">= 6"
   s.add_dependency "rouge"
-  s.add_dependency "sprockets", "< 4"
+  s.add_dependency "sprockets"
```
Comment (kevindew) [inline govuk_publishing_components.gemspec:25]: We'd want this to specify a version number if we only support sprockets `>= 4`

File: spec/dummy/app/assets/config/manifest.js [hunk 4]
```diff
@@ -1,5 +1,6 @@
-
 //= link_tree ../images
 //= link_directory ../javascripts .js
 //= link_directory ../stylesheets .css
-//= link govuk_publishing_components_manifest.js
+//= link component_guide/application.js
+//= link component_guide/application.css
```
Comment (kevindew) [inline spec/dummy/app/assets/config/manifest.js:5]: We should be able to have this as part of the govuk_publishing_components manifest

**Comment by alex-ju** [comment]
> Hey @alex-ju, as discussed on Slack, I've also been working on this - and I'm sorry we've ended up both putting energy into the same thing (I wonder what force of nature inspired us both.
> 
> The branch I've got is: https://github.com/alphagov/govuk_publishing_components/tree/sprockets-4-kevin and you can see the diff in: https://github.com/alphagov/govuk_publishing_components/compare/sprockets-4-kevin
> 
> Probably the major difference between our approaches is that I took a stance of supporting both Sprockets 3 and 4 for a while and warning users to upgrade.
> 
> There's a few things that seem to be working in your branch that surprise me too. I'll leave some comments surrounding those.
> 
> I'm not precious about my work on this so happy for you to make my bits or to open my own PR. Let me know your preference.

A backwards compatible approach would be ideal – I couldn't manage to make it work. Considering you already have a branch up and that I won't be around for long (my recent PRs in this repo took on average a month to get merged) I'd rather retire this in favour of yours. It was a good experiment, but requires a breaking change and a mass update.

**Pull request closed without merging**